src/transformers/models/llava_next/modeling_llava_next.py

```diff
@@ -404,29 +591,84 @@ def _merge_input_ids_with_image_features(self, image_features, inputs_embeds, in
             final_labels[batch_indices, text_to_overwrite] = labels[batch_indices, non_image_indices]
 
         # 5. Fill the embeddings corresponding to the images. Anything that is still zeros needs filling
-        image_to_overwrite = torch.all(final_embedding == 0, dim=-1)
-        image_to_overwrite &= image_to_overwrite.cumsum(-1) - 1 >= nb_image_pad[:, None].to(target_device)
-
-        if image_to_overwrite.sum() != image_features.shape[:-1].numel():
-            raise ValueError(
-                f"The input provided to the model are wrong. The number of image tokens is {torch.sum(special_image_token_mask)} while"
-                f" the number of image given to the model is {num_images}. This prevents correct indexing and breaks batch generation."
-            )
-
-        final_embedding[image_to_overwrite] = image_features.contiguous().reshape(-1, embed_dim).to(target_device)
+        with torch.no_grad():
+            image_to_overwrite = torch.all(final_embedding == 0, dim=-1)
+            if left_padding:
+                # exclude padding on the left
+                val = (
+                    max_embed_dim
+                    - torch.arange(max_embed_dim).unsqueeze(0).to(target_device).expand(batch_size, max_embed_dim)
+                ) <= embed_sequence_lengths[:, None].to(target_device)
+                image_to_overwrite &= val
+            else:
+                # exclude padding on the right
+                val = torch.arange(max_embed_dim).unsqueeze(0).to(target_device).expand(
+                    batch_size, max_embed_dim
+                ) < embed_sequence_lengths[:, None].to(target_device)
+                image_to_overwrite &= val
```

Comment:
We can rewrite to make a bit clearer

```suggestion
            embed_indices = torch.arange(max_embed_dim).unsqueeze(0).to(target_device)
            embed_indices = embedding_indices.expand(batch_size, max_embed_dim)
            embed_seq_lens = embed_sequence_lengths[:, None].to(target_device)

            if left_padding:
                # exclude padding on the left
                val = (max_embed_dim - embed_indices) <= embed_seq_lens
            else:
                # exclude padding on the right
                val = embed_indices < embed_seq_lens
            image_to_overwrite &= val
```